rr_platform/src/imu/imu.cpp

```diff
@@ -0,0 +1,139 @@
+
+void parse_message(const std::string &line_in, sensor_msgs::Imu &imu_msg, sensor_msgs::MagneticField &mag_msg, rr_platform::axes &axes_msg, bool &imu_ready, bool &mag_ready, bool &axes_ready){
+
+    imu_ready = false;
+    mag_ready = false;
+    axes_ready = false;
+
+
+    if (line_in.empty()) {
+        return;
+    }
+
+    std::vector <std::string> data = split(line_in, ',');
+
+    if(data[0] == "ax"){
+        imu_msg.linear_acceleration.x = std::atof(data[1].c_str());
+        imu_msg.linear_acceleration.y = std::atof(data[2].c_str());
+        imu_msg.linear_acceleration.z = std::atof(data[3].c_str());
+    }
+    else if(data[0] == "gx") {
+        imu_msg.angular_velocity.x = std::atof(data[1].c_str());
+        imu_msg.angular_velocity.y = std::atof(data[2].c_str());
+        imu_msg.angular_velocity.z = std::atof(data[3].c_str());
+    }
+    else if(data[0] == "q0") {
+        imu_msg.orientation.x = std::atof(data[1].c_str());
+        imu_msg.orientation.y =std::atof(data[2].c_str());
+        imu_msg.orientation.z = std::atof(data[3].c_str());
+        imu_msg.orientation.w = std::atof(data[4].c_str());
+        imu_ready = true;
+    }
+    else if(data[0] == "mx") {
+        mag_msg.magnetic_field.x = std::atof(data[1].c_str());
+        mag_msg.magnetic_field.y = std::atof(data[2].c_str());
+        mag_msg.magnetic_field.z = std::atof(data[3].c_str());
+        mag_ready = true;
+    }
+    else if(data[0] == "axes") {
+        axes_msg.roll = std::atof(data[1].c_str());
+        axes_msg.pitch = std::atof(data[2].c_str());
+        axes_msg.yaw = std::atof(data[3].c_str());
+        axes_ready = true;
+    }
+}
+
+
+int main(int argc, char **argv) {
+
+    ros::init(argc, argv, "imu");
+
+    ros::NodeHandle handle;
+    ros::NodeHandle private_handle("~");
+
+    imu_pub = handle.advertise<sensor_msgs::Imu>("/imu/data_raw", 1);
+    mag_pub = handle.advertise<sensor_msgs::MagneticField>("/imu/mag", 1);
+    axes_pub = handle.advertise<rr_platform::axes>("/axes", 1);
+
+    // Serial port setup
+    std::string serial_port_name;
+    private_handle.param(std::string("serial_port"), serial_port_name, std::string("/dev/ttyACM0"));
+    SerialPort serial_port;
+    
+    if(!serial_port.Open(serial_port_name, 115200)) {
+        ROS_FATAL_STREAM("Unable to open serial port: " << serial_port_name);
+        return 1;
+    }
+
+
+    sensor_msgs::Imu imu_msg;
+    sensor_msgs::MagneticField mag_msg;
+    rr_platform::axes axes_msg;
+
+    imu_msg.header.frame_id = "imu";
+    mag_msg.header.frame_id = "mag";
+
+    bool imu_ready = false;
+    bool mag_ready = false;
+    bool axes_ready = false;
+
+
+    ROS_INFO("IMU Ready.");
+
+    ros::Duration(2.0).sleep();
+
+    while(ros::ok()) {
+        ros::spinOnce();
+
+        auto line_in = serial_port.ReadLine();
+
+        if(!line_in.empty()) {
+
+            parse_message(line_in, imu_msg, mag_msg, axes_msg, imu_ready, mag_ready, axes_ready);
+
+            if(imu_ready) {
```

Comment:
This whole ready flag system is very unnatural for ROS. 
I'd recommend breaking up the different parsing blocks into their own functions. Each one returns a pair with the message and a success bool. Then this main loop can simply ask each function to look for its message type in the line, and if it finds it, this loop will publish it.
This idea will also clean up very nicely when we can use C++17's if statements with initializers.